---
layout: preprint
#layout: review, 11pt
title: "RNeXML: Parsing and Serializing the Next Generation of Phyloinformatic Data in R"
author: 
  - name: Carl Boettiger
    affiliation: cstar
    email: cboettig@gmail.com
    footnote: Corresponding author
  - name: Hilmar Lapp
    affiliation: NESCent
  - name: Scott Chamberlain
    affiliation: SimonFraser
  - name: Rutger Vos
    affiliation: NCB
address: 
  - code: cstar
    address: | 
      Center for Stock Assessment Research, 
      Department of Applied Math and Statistics, 
      University of California, Mail Stop SOE-2,
      Santa Cruz, CA 95064, USA
  - code: NESCent
    address: | 
      National Evolutionary Synthesis Center, Durham, NC, USA
  - code: SimonFraser
    address: |
      Department of Biology, Simon Fraser University, Burnaby, Canada
  - code: NCB
    address: | 
      NCB Naturalis, Leiden, the Netherlands
abstract: | 
      1. NeXML promises to be the next generation of phylogenetic
      informatics data and metadata exchange format, offering substantial
      improvements in reliability, extensibility, and richness over
      earlier standards such as Nexus.

      2. In this paper we present the RNeXML package which provides
      a bridge between this rich and reliable data format and the
      extensive library of phylogenetic tools already available in R,
      with a particular emphasis on comparative methods. 

      3. RNeXML allows R users to both read and write NeXML files. Yet
      RNeXML is much richer than a basic parser: here we illustrate
      how the package facilitates adding and manipulating semantic 
      annotations.

      4. An essential feature of NeXML is to provide both the flexibility
      to be continually extended to meet the emerging needs of the phylogenetics
      community while remaining backwards-compatible with software developed
      for earlier versions.  Here we illustrate how RNeXML makes it easy
      to extend NeXML in this way.  
      
      5. RNeXML is perfectly suited for users to preserve and maintain their
      personal archives of phylogenetic and comparative trait data in consistent,
      query-able, and cross-platform format, or leverage the rapidly growing 
      public community archives of rich NeXML files.  



---

<!-- Emphasize: 
1) storing character data and tree data together, convenient for the comparative methods community
2) Adding metadata, including automated addition of metadata and semantic utilities 
3) Extending the NeXML standard  

-->

```{r compile-settings, include=FALSE}
library(knitr)
opts_chunk$set(tidy = FALSE, warning = FALSE, message = FALSE, 
               cache = 1, comment = NA, verbose = TRUE)
basename <- gsub(".Rmd", "", knitr:::knit_concord$get('infile') )
opts_chunk$set(fig.path = paste("figure/", basename, "-", sep=""),
               cache.path = paste("cache/", basename, "/", sep=""))


```


Though R provides the most extensive and widely used collection of
comparative phylogenetic methods tools, R users lack a convenient,
extensible format to exchange and archive comparative methods data.


## Background

### Challenges with existing formats
See [Vos _et al._ 2012].  

### Challenges with R formats

- Most R users rely on the nexus inherits the limitations of nexus file format

See [phylobase vignette, 2010]


<!--
A summar of some phylogenetic formats in R:

| Class     | R package   | R Object 
|-----------|-------------|-------
| phylo     | ape         | S3 
| phylo4    | phylobase   | S4
| phylo4d   | adephylo    | S4
| ouchtree  | ouch        | S4 
| treeshape | apTreeshape | list 
-->

### NeXML

See [Vos _et al._ 2012].

Mention of data sharing challenges and opportunities.  See [Drew et al 2013]




Why RNeXML?

* Why NeXML -- because Nexus standard isn't a standard.  Developers extend it arbitrarily to meet their needs, making it incompatible <!-- with? -->.  For example, Nexus cannot represent horizontal gene transfer, ambiguous use of symbols, cannot be extended reliably, and there is no way to validate that it is proper Nexus.  NeXML addresses all of these problems [Vos _et al._ 2012] with a state-of-the-art data management format.  

* Why RNeXML: Huge ecosystem of phylogenetics packages. (1) Read NeXML.
All would benefit by being able to read NeXML, both as a robust standard
free of the pitfalls of Nexus, and more importantly, as a much richer
source of data thanks to its extendable annotation model.  (2) Write
NeXML.  R packages and R users add rich information to the phylogenies
they work with, but need a platform independent data exchange format to
share and archive this. In particular, we illustrate how NeXML can
be used to distribute both phylogenies and character trait data used in 
many popular phylogenetic comparative methods tools in R.  (3) Annotate 
and extend nexml. `RNeXML` provides convenient tools for adding metadata
to NeXML files, including the automated generation of metadata and examples
of how to extend the NeXML standard to represent new data features without
breaking validation or backwards compatability of the file type.  



## Introduction to RNeXML


A nexml file can be thought of as having three components: (1) phylogeny
data (2) character-trait data, and (3) metadata. `RNeXML` seeks to
provide a seemless way to convert these to and from their corresponding
R object representations, such as `ape::phylo` phylogenies, character
matrices (`data.frames`) and metadata lists (or for advanced users,
`XMLInternalDocument` and `rrdf` triplestores). Metadata can provide
information about the study in which the phylogeny was generated or
applied (such as authors, citations to publications, timestamps, and even
descriptions of methods), but can also be used to annotate individual
data components.  We illustrate how metadata annotations can be added to
the operational taxonomic units of the phylogeny and can be used to extend
the data representation to new uses such as stochastic character maps.

The current stable version of the `RNeXML` package can be installed from
the CRAN library using the standard installation method <!-- Actually not yet --> 

```r
install.packages("RNeXML")
```

The latest stable development version can be installed from Github using the `devtools` package:

```r
library(devtools)
install_github("RNeXML", "ropensci")
```


## Working with phylogenetic trees

#### Read methods

```{r echo=FALSE}
library(RNeXML)
```

We begin by reading in an example NeXML file provided with the package.  

```{r}
f <- system.file("examples", "trees.xml", package="RNeXML")
nex <- nexml_read(f)
```

(For convenience `nexml_read` is also aliased as `read.nexml`, for
users familiar with the somewhat analogous `read.nexus` function).
The resulting `nex` object is an R object representation of the NeXML.
From this object we can extract any phylogenies it contains in the
`ape::phylo` format:

```{r}
phy <- get_trees(nex)
```

We can then leverage the rich suite of methods available from the R
phylogenetics community (including coercion into other representations,
such as `ouch` [@ouch] and `phylobase` [@phylobase] phylogeny
formats). For instance, here we use the standard plotting function from
the `ape` package to plot the resulting `ape::phylo` object returned by
`get_trees`.



```{r eval=FALSE}
plot(phy)
```

NeXML can not only contain multiple phylogenetic trees in a single file,
but can also group sets of trees into separate blocks.  The `get_trees`
function attempts to return the most sensible R object for NeXML: if
only a single tree is available, it will return an `ape::phylo` object;
if multiple trees are available in a single `trees` block, it will return
an `ape::multiPhylo` object (a list of `phylo` objects).  If multiple
blocks are available, it will return a list of `multiPhylo` blocks.
While most users will find this behavior convenient, there are times
(such as looping over many NeXML files) when having a variable return
type is undisirable.  To always get back a list of `multiPhylo` objects,
even when only one tree is available, use `get_trees_list()`.

<!-- have we already said all this above?-->  <!-- no, seems appropriate to repeat here, -scott --> The ability to read in
phylogenetic trees in the `nexml` format opens up a wide and rapidly
growing array of resources to the R user.  For instance, a user can access
all the phylogenies available in TreeBASE through the `nexml` format.
`RNeXML` can read directly from a URL:

```{r eval=FALSE}
nex = nexml_read("https://raw.github.com/rvosa/supertreebase/master/data/treebase/S100.xml")
```

Previously, this was possible using the `treebase` package [@treebase]
for R, which relied on the nexus parser provided in the `ape` package.
Due to inconsistencies present in the Nexus standard [Vos _et al._
2012], a significant number <!-- State explicitly --> of TreeBASE nexus
files could not be read in this way.  Using the more precisely defined
NEXML standard, every valid NeXML file on TreeBASE can now successfully
be parsed to R objects.



#### Character data 

Like nexus files, NeXML files can also contain character data.  While
most researchers will be familiar with the use of character matrices in
nexus files as a container for molecular sequence data used to infer
the phylogenetic tree, we find this ideal for character data used in
comparative methods popular in R. Methods to get extract character data
work much like those for the phylogenetic data:

```{r }
comp_analysis <- system.file("examples", "comp_analysis.xml", package="RNeXML")
nex <- nexml_read(comp_analysis)
get_characters(nex)
```

Returns a `data.frame` with columns as characters and rows as
taxa. Continuous and discrete characters are represented as seperate
blocks in NeXML files, but will be combined as seperate columns of a
single data frame by `get_characters()` if they correspond to the same
taxa. The `get_characters_list()` function will consistently recieve
a list of data frames corresponding to each character matrix block of
the NeXML.


#### Write methods 

Generating NeXML files from R is likewise straight forward.  Here we
write a phylogeny from the `ape::phylo` format out to a NeXML file:

```{r}
data(bird.orders)
nexml_write(bird.orders, file = "birds.xml")
```

The first argument to `nexml_write` can be a `phylo` phylogeny,
`multiPhylo` list of phylogenies, or a `nexml` object -- anything that
can be coerced into the `nexml` class.  As we have seen, NeXML files
can contain multiple phylogenetic trees, as well as character matrices.
We can add both trees  and character data using `nexml_write`,

```{r}
library(geiger)
data(geospiza)
nexml_write(trees = geospiza$phy, characters = geospiza$dat, file="geospiza.xml")
```


NeXML files can also be built up bit by bit using the `add_characters`,
`add_trees`, and `add_meta`.  Each of these functions can add data to an
existing `nexml` object or create a new nexml object if none is provided.
Here we first create a `nexml` object containing the phylogeny data,
and then add the character data to it:

```{r}
geiger_nex <- add_trees(geospiza$phy)
geiger_nex <- add_characters(geospiza$dat, geiger_nex)
```

The data need not share the same taxa.  We can append additional
phylogenies and characters corresponding to different taxa:

```{r}
data(primates)
geiger_nex <- add_trees(primates$phy, geiger_nex)
geiger_nex <- add_characters(primates$dat, geiger_nex)
```

This allows us to include multiple trees and character matrices to a single NeXML file.  
<!-- We can add relevant metadata to each tree, such as the citation information 
     indicating the paper in which it was first described.  -->


### Metadata

At the heart of the NeXML standard is the ability to encode arbitrary
metadata in precisely specified, machine-readable formats that maintain
backwards compabitibilty with any existing NeXML parser. NeXML uses the
"Resource Description Framework in Annotations", or RDFa [W3C 2013],
as described in [Vos et al 2011].  These annotations can be used to add 
information to any particular element of the data, (a tree, node, edge,
or taxonomic unit; a character matrix or specific trait, or even another 
metadata annotation or the NeXML data file as a whole). `RNeXML` provides
a range of tools for accessing metadata embedded in a NeXML file.  



### Writing NeXML metadata

Researchers cannot take advantage of NeXML metadata manipulation if
there is no metadata there.  The burden of adding good metadata remains a
significant barrier to increased sharing and reuse of scientific research
[Tenopir et al. 2011].  `RNeXML` seeks to reduce the burden of adding
high quality, machine-readable metadata to NeXML files. This occurs on
several levels:

- Automatic annotation.  

`RNeXML` automatically adds basic metadata by default such as a timestamp
of when the tree was produced and an open data license (CC0, compatible
with deposition on Dryad data repository), which can be configured
as needed.  To adjust these and other common metadata fields that should 
be part of any data deposition, `RNeXML` provides the `add_basic_meta()`
function.  This function does not assume any familiarity with the concepts
of namespaces or other more advanced metadata tools discussed later.  The
`add_basic_meta` takes an existing `nexml` object (or will create a new one
if none is provided, like the other `add_` functions), and any of the following
fields: `title`, `description`, `creator`, `pubdate`, `rights`, `publisher`,
`citation`.  Additional fields may be added to subsequent versions, see
`?add_basic_metadata` for the most up-to-date information.  Here we create
an new `nexml` object with the basic metadata for the `bird.orders` data
provided in the `ape` package:

```{r}
data(bird.orders)
birds <- add_trees(bird.orders)
birds <- add_basic_meta(birds,
  title = "Phylogeny of the Orders of Birds From Sibley and Ahlquist",

  description =" This data set describes the phylogenetic relationships of the
     orders of birds as reported by Sibley and Ahlquist (1990). Sibley
     and Ahlquist inferred this phylogeny from an extensive number of
     DNA/DNA hybridization experiments. The ``tapestry'' reported by
     these two authors (more than 1000 species out of the ca. 9000
     extant bird species) generated a lot of debates.

     The present tree is based on the relationships among orders. The
     branch lengths were calculated from the values of Delta T50H as
     found in Sibley and Ahlquist (1990, fig. 353).",

  citation = "Sibley, C. G. and Ahlquist, J. E. (1990) Phylogeny and
     classification of birds: a study in molecular evolution. New
     Haven: Yale University Press.",

  creator = "Sibley, C. G. and Ahlquist, J. E.")
```

Citations can also be added using R's `bibentry` type, which allows
a user to add citations to R packages, 

```{r}
birds <- add_basic_meta(birds, citation = citation("ape"))
```

or to papers by simply using the object's DOI:

```{r}
library("knitcitations")
geiger_nex <- add_basic_meta(geiger_nex, citation=cite("10.2307/2408428"))
```

(Which is the paper on which the `geospiza` phylogeny from the geiger dataset is based)


`RNeXML` also provides various functions to assist in automatically adding
other more specific metadata to various elements of the NeXML data.  



For instance, misspellings or inconsistent use of taxonomic names in phylogenetic
data are a common source of frustration in reusing phylogenetic trees. By 
associating the taxonomic labels against a specific, centralized database
we can catch typographical errors and indicate more precisely which classification
we are appealing to in the event of conflicting taxonomic definitions.  In this 
example, we indicate that we intend the bird orders listed in the tree to correspond
to the NCBI definitions 

```{r message=FALSE, results='hide'}
birds <- taxize_nexml(birds, "NCBI")
```

This function uses the `taxize` R library ([Chamberlain 2013]) to check each taxon
label against the NCBI database.  If a unique match is found, a metadata annotation
is added to the taxon providing the NCBI identification number to the taxonomic unit.
If no match is found, the user is warned to check for possible typographic errors in
the taxonomic labels provided.  

### Custom metadata extensions

So far our examples have relied on functions that have a built-in understanding of the 
kind of metadata being provided, and can thus generate the appropriate metadata with 
little further guidence.  In order to be machine-readable and understandable, 
NeXML requires all metadata annotations follow the RDFa format which explicitly declares 
the meaning of the annotation in a way a computer can understand.  For instance, the 
title we gave to the `bird.orders` phylogeny appears in the NeXML as:

```xml
<meta id="m2" property="dc:title" datatype="xsd:string"
      content="Phylogeny of the Orders of Birds From Sibley and Ahlquist"/> 
```

We see the title appears as the `content` of this `meta` element.  The `id` attribute
provides us with a unique way to refer to this meta element in this document.  The `datatype`
tells us that the content is a text string (as opposed to a numerical or boolean value).  
The `property` is the attribute that most interests us here; it tells us that the content 
is a title.  The prefix `dc` indicates the vocabulary (called a namespace in XML) to which 
the word `title` belongs.  The prefix is defined elsewhere in the NeXML file, as we can see with:


```{r}
prefixes <- get_namespaces(birds)
prefixes["dc"]
```

This URL links us to a webpage providing both a human and machine-readable version of precisely
what the term `title` means, showing us that it refers to `title` in the sense of a creative work, 
rather than the title of a person, a legal title, the verb title, etc.  This precision can allow
us to perform precise computer reasoning on metadata, without relying on heuristics or assumptions,
as we illustrate later.  For the moment though, the need to refer to an explicit controlled vocabulary (or
ontology) with a url, saying "dc:title" rather than just saying the property is "title", is something
of a burden in creating metadata.  The previous examples involve functions that already know what 
kind of metadata is being provided (a title, a data creator, etc), and so can provide the appropriate
vocabulary automatically.  The main advantage of linking to an external vocabulary, however, is the
ability to express concepts that are not already "built in" to the data model.  `RNeXML` provides 
tools to generate arbitrary metadata annotations given an external vocabulary, such as we illustrate
here.  


From the list of namespaces in `prefixes`, we can follow links a handful of established vocabularies
and find terms that are not included in `add_basic_meta`.  For instance, in the `prism` prefix, 

```{r}
prefixes["prism"]
```

we see that there is a term for `modificationDate` for a file.  We create a `meta` element 
containing this annotation using the `meta` function:

```{r}
modified <- meta(property = "prism:modificationDate", content = "2013-10-04")
```

We can add this annotation to our existing `birds` nexml file using the `add_meta` function.
Because we do not specify a level, it is added to the root node, refering to the nexml file as a whole.  

```{r}
birds <- add_meta(modified, birds) 
```

The built-in vocabularies are just the tip of the iceburg of established vocabularies. Here 
we add an annotation from the `skos` namespace which describes the history of where the
data comes from:

```{r}
history <- meta(property = "skos:historyNote",
  content = "Mapped from the bird.orders data in the ape package using RNeXML")
```

Because `skos` is not in the current namespace list, we add it with a url when
adding this meta element.  We also specify that this annotation be placed 
at the level of the `trees` subnode in the NeXML file.  

```{r}
birds <- add_meta(history, 
                birds, 
                level = "trees",
                namespaces = c(skos = "http://www.w3.org/2004/02/skos/core#"))
```

If we forget to add the namespace with add_meta, we can always add it later
using the `add_namespaces` function,


```{r}
birds <- add_namespaces(c(skos = "http://www.w3.org/2004/02/skos/core#"), birds)
```

(The `add_namespaces` function will avoid creating any duplicate definitions)
For finer control of the level at which a `meta` element is added, we will manipulate 
the nexml R object directly using S4 subsetting, as shown later.  


<!-- Ugh. y'all have much more practice explaining this well and without tangents, 
so please feel free to edit or re-write.  -->

Much richer metadata annotation is possible. Later we illustrate how metadata annotation
can be used to extend the base NeXML format to represent new forms of data while maintaining
compatability with any NeXML parser. The `RNeXML` package can be easily extended to support
helper functions such as `taxize_nexml` to add additional metadata without imposing a large
burden on the user.  

<!--
- Utilities to add common metadata

  - Add a citation to any papers associated with the phylogeny data.  
  - Writing in methodological information on how tree has been generated or analyzed (including R code)
  - Add annotations to any taxon.  

Temporal, spatial, taxonomic coverage for dryad/knb?
--> 



### Reading NeXML metadata

A call to the nexml object prints some metadata summarizing the data structure: <!-- need to make this method faster!--> 


```{r }
birds
```

We can extract all metadata pertaining to the NeXML document as a whole
(annotations of the XML root node, `<nexml>`) with the command

```{r}
meta <- get_metadata(birds) 
```

This returns a named list of available metadata. We can see the kinds
of metadata recorded from the names (showing the first 4):

```{r}
names(meta)[1:4]
```

and can ask for a particular element using the standard list subsetting
mechanism (i.e. either the name of an element or it's numeric position),

```{r}
meta[["dc:title"]]
```

All metadata terms must belong to an explicit *namespace* or vocabulary
that allows a computer to interpret the term precisely. The prefix (before
the `:`) indicates to which vocabulary the term belongs, e.g. `dc` in
this case.  The `get_namespaces` function tells us the definition of
the vocabulary using a link:

```{r}
prefixes <- get_namespaces(birds)
prefixes["dc"]
```

Following the link we can see that `dc` refers to the "Dublin Core"
vocabulary, and the term `title` refers to the title of a creative work
(as opposed to the title of a person, a legal title, the verb title, etc).
Such explicit definitions allow a computer to reason about these objects,
as we shall see later.  


Common metadata can be accessed with a few dedicated functions:

```{r}
get_citation(birds)
```


```{r}
get_taxa(birds)
```

Which returns text from the otu element labels, typically used to define
taxonomic names, rather than text from explicit meta elements.



We can also access metadata at a specific level (or use `level=all`
to extract all meta elements in a list).  Here we show only the first
few results:

```{r} 
otu_meta <- get_metadata(birds, level="otu")
otu_meta[1:4]
```

This approach does not preserve the relationship between the metadata element and the 
`otu` it was annotating, and as such is most useful for quick visual exploration. 
Metadata values for specific elements or sets of elemetns can always be accessed using
the S4 subsetting mechanisms (see section below).  



## RDF manipulation and SPARQL queries 

Consider a better example than this one: 


```{r}
rdf <- get_rdf(birds)
```

Query the rdf using XPath expressions:

```{r}
library(XML)
xpathSApply(rdf, "//dc:title", xmlValue) 
```

XPath is an expressive syntax for querying XML-structured data such
as RDF.  Queries using the XPath syntax are possible in R using the
XML library [Temple Lang 2013, XML].  (Note that XPath is far more
expressive for querying the metadata once the RDFa has been extracted,
since attribute values are now interpreted as namespaced XML nodes and
not just arbitrary text.)

The most powerful way to take advantage of the machine-readable nature
of the metadata involves the construction of SPARQL queries. This
approach can exploit the reasoning and logical deduction capacity of
formal vocabularies (ontologies). For a proper introduction to SPARQL
queries, see CITE SOMETHING.  Meanwhile, we just introduce the sytnax
necessary to apply such queries to NeXML files from R.  First we load
the necessary R library and import the RDF-extracted metadata:

```{r}
library(rrdf)
saveXML(rdf, "rdf_meta.xml") # rrdf requires a file name, so we must write the XML out first
lib <- load.rdf("rdf_meta.xml")
```

We can then make arbitrary SPARQL queries directly from R.  For instance,
here is a simple query to extract the value of any `dc:title` element, analogous
to the XPath expression above: 

```{r}
sparql.rdf(lib, "SELECT ?title WHERE { ?x <http://purl.org/dc/elements/1.1/title> ?title}")
```




### Extending the NeXML standard through metadata annotation.

No data standard can anticipate all the uses and extensions researchers
may one day need. Researchers have repeatedly adapted the existing NEXUS
file format to accomidate encoding new kinds of information, but in the
process break compatibility between different flavors of NEXUS files, with
few reliable methods to distinguish between these flavors.  NeXML provides
a solution to this problem through the use of metadata annotation.

Here we illustrate this process using the example of stochastic character
mapping [Huelsenbeck et al 2003]. A stochastic character map is simply
an annotation of the branches on a phylogeny, assigning each section
of each branch to a particular "state" (typically of a morphological
characteristic).

[Bollback 2006] provides a widely used stand-alone software implementation
of this method in the software `simmap`, which modified the standard
Newick tree format to express this additional information. This can
break compatibility with other software[^1], and creates a format that
cannot be interpreted without additional information describing this
convention.  By contrast, the NeXML extension is not only backwards
compatible but contains a precise and machine-readable description of
what it is encoding.

In this example, we illustrate how the additional information required
to define a stochastic character mapping (a `simmap` mapping) in NeXML.  


[^1]: By using the commenting mechanism of the Newick format, it is
possible that other software that doesn't also use the comment mechanism
for some other such purpose would be able to successfully parse the tree.
However there is no way to guarentee that this is the case or for the
data format to describe its use.

[Revell 2011] describes the `phytools` package for R, which includes
utilities for reading, manipulating, and writing `simmap` files in R.
In this example, we also show how to define `RNeXML` functions that 
map the R representation used by Revell (an extension of the `ape` class)
into the NeXML extension we have defined by using `RNeXML` functions..

Since a stochastic character map simply assigns different states to
parts of a branch (or edge) on the phylogenetic tree, we can create
a NeXML representation by annotating the `edge` elements with appropriate
`meta` elements.  These elements need to describe the character state
being assigned and the duration (in terms of branch-length) that the edge 
spends in that state (Stochastic character maps are specific to time-calibrated
or ultrametric trees).  

NeXML already defines the `characters` element to handle descrete character traits (`nex:char`)
and the states they can assume (`nex:state`).  We will thus reuse the `characters` element for
this purpose, referring to both the character trait and the states by the ids assigned to them
in that element.  (NeXML's convention of referring to everything by id permits a single canonical
definition of each term, making it clear where additional annotation belongs).  For each edge, we 
need to indicate:

<!-- Ugh, do we need to explain that everything in NeXML uses an id so that it can be reused in this way?
     -->

- That our annotation contains a stochastic character mapping reconstruction
- Since many reconstructions are possible for a single edge, we give each reconstruction an id
- We indicate for which character trait we are defining the reconstruction 
- We then indicate which states the character assumes on that edge. 
  For each state realized on the edge, that involves stating: 
    + the state assignment
    + the duration (length of time) for which the edge spends in the given state
    + the order in which the state changes happen (Though we could just assume 
      state transitions are listed chronologically, NeXML suggests making all 
      data explicit, rather than relying on the structure of the data file to
      convey information).  

Thus the the annotation for an edge that switches from state `s2` to state 
`s1` of character `cr1` would be constructed like this:

```{r}
 m <- meta("simmap:reconstructions", children = c(
        meta("simmap:reconstruction", children = c(

          meta("simmap:char", "cr1"),
          meta("simmap:stateChange", children = c(
            meta("simmap:order", 1),
            meta("simmap:length", "0.2030"),
            meta("simmap:state", "s2"))),
          
          meta("simmap:char", "cr1"),
          meta("simmap:stateChange", children = c(
            meta("simmap:order", 2),
            meta("simmap:length", "0.0022"),
            meta("simmap:state", "s1")))
          ))))
```

Of course writing out such a definition manually becomes tedious quickly. Because
these are just R commands, we can easily define a function that can loop over an
assignment like this for each edge, extracting the appropriate order, length and
state from an existing R object such as that provided in the `phytools` package.  
Likewise, it is straight forward to define a function that reads this data using
the `RNeXML` utilities and converts it back to the `phytools` package. The full
implementation of this mapping can be seen in the `simmap_to_nexml()` and the
`nexml_to_simmap()` functions provided in the `RNeXML` package.  

As the code indicates, the key step is simply to define the data in meta elements. In 
so doing, we have defined a custom namespace, `simmap`, to hold our variables.  This
allows us to provide a URL with more detailed descriptions of what each of these 
elements mean:

```{r}
nex <- add_namespaces(c(simmap = "https://github.com/ropensci/RNeXML/tree/master/inst/simmap.md"))
```

At that URL we post a simple description of each term. 





[Revell 2011]: http://doi.org/10.1111/j.2041-210X.2011.00169.x
[Huelsenbeck et al 2003]: http://doi.org/10.1080/10635150390192780 "Stochastic Mapping of Morphological Characters"
[Bollback 2006]: http://doi.org/10.1186/1471-2105-7-88 "SIMMAP: stochastic character mapping of discrete traits on phylogenies."





### Subsetting nexml objects directly

The `RNeXML` package provides many convenient functions to add and extract
information from `nexml` objects in the R environment without requiring
the reader to understand the details of the NeXML data structure and
making it less likely that a user will generate invalid NeXML syntax
that could not be read by other parsers. The `nexml` object we have been using
in all of the examples is built on R's S4 mechanism. Advanced users may
sometimes prefer to interact with the data structure more directly using 
R's S4 class mechanism and subsetting methods. Many R users are more familiar
with the S3 class mechanism (such as in the `ape` package phylo objects)
rather than the S4 class mechanism used in phylogenetics packages such as 
`ouch` and `phylobase`. The `phylobase` vignette provides an excellent introduction
to these data structures.  Users already familiar with subsetting lists and other
S3 objects in R are likely familar with the use of the `$` operator, such as
`phy$edge`. S4 objects simply use an `@` operator instead (but cannot be subset
using numeric arguments such as `phy[[1]]` or named arguments such as phy[["edge"]]).  


The `nexml` object is an S4 object, as are all of its components (slots).  It's 
heirachical structure corresponds exactly with the XML tree of a NeXML file, with 
the single exception that both XML attributes and children are represented as slots.  
S4 objects have constructor functions to initialize them.  We create a new `nexml` 
object with the command:

```{r}
nex <- new("nexml")
```

We can see a list of slots contained in this object with

```{r}
slotNames(nex)
```

Some of these slots have already been populated for us, for instance, the schema version and default namespaces:

```{r}
nex@version
nex@namespaces
```

Recognize that `nex@namespaces` serves the same role as `get_namespaces` function, but provides direct access to the slot data.  For instance, with this syntax we could also overwrite the existing namespaces with `nex@namespaces <- NULL`.  Changing the namespace in this way is not advised.

Some slots can contain multiple elements of the same type, such as `trees`, `characters`, and `otus`.  For instance, we see that 

```{r}
class(nex@characters)
```

is an object of class `ListOfcharacters`, and is currently empty,

```{r}
length(nex@characters)
```

In order to assign an object to a slot, it must match the class definition
of the slot.  We can create a new element of any given class with the
`new` function,

```{r}
nex@characters <- new("ListOfcharacters", list(new("characters")))
```

and now we have a length-1 list of character matrices,

```{r}
length(nex@characters)
```

and we access the first character matrix using the list notation, `[[1]]`. Here we check the class is a `characters` object.  

```{r}
class(nex@characters[[1]])
```

Direct subsetting has two primary use cases: (a) useful in looking up (and possibly editing) a specific value of an element, or (b) when adding metadata annotations to specific elements. Consider the example file


```{r}
f <- system.file("examples", "trees.xml", package="RNeXML")
nex <- nexml_read(f)
```

We can look up the species label of the first `otu` in the first `otus` block:

```{r}
nex@otus[[1]]@otu[[1]]@label
```

We can add metadata to this particular OTU using this subsetting format

```{r}
nex@otus[[1]]@otu[[1]]@meta <- 
  c(meta("skos:note", 
          "This species was incorrectly identified"),
         nex@otus[[1]]@otu[[1]]@meta)
```

Here we use the `c` operator to append this element to any existing meta annotations to this otu.  




### Publishing NeXML files from R

Data archiving is increasingly required by journals in evolutionary biology and 
biodiversity [e.g. @Whitlock2011].  The burden of preparing and submitting properly
annotated data to archives continues to be a significant barrier [Tenopir et al. 2011, Stodden 2013],
and many phylogenetic trees upon which studies are based are inaccessible or lost 
to the research community [Drew et al 2013].  `RNeXML` seeks to lower these barriers
by providing functions to immediately archive NeXML files in scientific repositories.  




```{r eval = FALSE}
doi <- nexml_publish(birds, visibility = "public", repository="figshare")
```


_Show downloading the nexml file directly from figshare, explain value
of global paths instead of personal directory paths appearing in scripts, etc_



Robust data archiving need not involve immediate release of the data. The `figshare` 
repository supports the option of `visibility = "private"`, allowing secure 
publication that provides a way to backup and share data with collaborators 
prior to public release.  

<!-- Wouldn't it be nice to publish to Dryad or TreeBASE in this way?  -->




### Implementation and architecture

The `RNeXML` package is built for the NeXML Schema, [version 2009],
using the R software environment and the XML package for parsing and
generating XML [Temple Lang 2013, XML], and provides conversions into
the popular `phylo` format provided by the `ape` package [Paradis 2004].
`RNeXML` can automatically check taxonomic names and add taxonomic
identifiers when generating NeXML using methods from the `taxize` package
([Chamberlain 2013]).

<!-- Possibly cite all dependencies and suggests from the DESCRIPTION --> 
<!-- Reference any installation and configuration instructions necessary (e.g. API credentials for figshare) -->


### Quality control

The `RNeXML` package performs severak levels of validation on NeXML files
written from or read into R: validating the XML, and validating against
the current NeXML schema, and performing additional checks through the 
NeXML online validation tool for elements that cannot be enforced by
schema validation alone.  Validation serves to improve consistency in
data encoding which facilitates the use of NeXML across a wide variety
of platforms and software.  A lack of a consistent standard and validation
mechanism has continually plagued the current NEXUS standard and made 
interoperability difficult [Vos et al. 2011].  

Several mechanisms are in place to evaluate the quality of the `RNeXML`
package functions themselves.  In addition to the examples from the
documentation and the automated package checks provided for the R system,
the `RNeXML` package includes a complete test suite using the `testthat`
package [Wickham 2011]. Package development on Github includes continuous 
integration through the `travis` platform, in which the full test suite
is run every time the code is changed.  This not only gives users confidence
that the development copy can be successfully installed, but should assist
other developers seeking to contribute to the package to test that their proposed 
changes have not negatively impacted any of the existing code base.  


### Software reuse and support

The `RNeXML` package is available through the CRAN repository under
a CC0 license. At the time of writing, the current version is `r packageDescription("RNeXML", fields="Version")`. 
The current development
branch can be found on Github at [https://github.com/ropensci/RNeXML].
See the NEWS file for changes in recent versions. Users are encouraged
to submit bug reports or feature requests in the issues log.  Contact
ropensci-discuss@googlegroups.com or the package maintainer for help.

There are many promising directions for further development of the
`RNeXML` package.  The project's [Milestones page][milestones] page provides a list of
steps currently planned or under consideration.  In particular, these
milestones include increased support for interactive mode of entering data
(wizards), and support for integration of semantics defined through the
Web Ontology language (OWL), as discussed in the Challenges section.



### Acknowledgements

This project was supported in part by a grant from the Alfred P Sloan
Foundation (CB and SC), NSF Grant DBI-1306697 (CB), ... `RNeXML` is part
of the rOpenSci project, [http://ropensci.org].



[Temple Lang 2013, XML]: http://cran.r-project.org/web/packages/XML/
[Wickham 2011]: http://vita.had.co.nz/papers/testthat.html
[Chamberlain 2013]: http://dx.doi.org/10.12688/f1000research.2-191.v2
[@ouch]: http://cran.r-project.org/web/packages/ouch/
[@phylobase]: http://cran.r-project.org/web/packages/phylobase
[W3C 2013]: http://www.w3.org/TR/xhtml-rdfa-primer/ "RDFa 1.1 Primer - Second Edition: Rich Structured Data Markup for Web Documents"
[Tenopir et al. 2011]: http://doi.org/10.1371/journal.pone.0021101 "Data Sharing by Scientists: Practices and Perceptions"
[Drew et al 2013]: http://doi.org/10.1371/journal.pbio.1001636 
[milestones]: https://github.com/ropensci/RNeXML/issues/milestones


phylobase::phylo4 was an attempt to solve the problems associated with
the popular ape::phylo format (unpredictable errors, loose and multiple
conflicting specifications).



