---
#layout: preprint
layout: review, 11pt
title: "RNeXML: a package for reading and writing richly annotated phylogenetic, character, and trait data in R"
author: 
  - name: Carl Boettiger
    affiliation: cstar
    email: cboettig(at)gmail.com
    footnote: Corresponding author
  - name: Hilmar Lapp
    affiliation: NESCent
  - name: Scott Chamberlain
    affiliation: ropensci
  - name: Rutger Vos
    affiliation: NBC
address: 
  - code: cstar
    address: | 
      Center for Stock Assessment Research, 
      Department of Applied Math and Statistics, 
      University of California, Mail Stop SOE-2,
      Santa Cruz, CA 95064, USA
  - code: NESCent
    address: | 
      National Evolutionary Synthesis Center, Durham, NC, USA
  - code: ropensci
    address: |
      University of California, Berkeley, CA, USA
  - code: NBC
    address: | 
      Naturalis Biodiversity Center, Leiden, the Netherlands
abstract: | 
      1. NeXML is a new exchange format for phylogenetic data and
      metadata that offers substantial improvements in reliability,
      extensibility, and richness over earlier formats such as
      NEXUS. In this paper we present the RNeXML package which
      provides a bridge between this data format and the extensive
      library of phylogenetic tools already available in the popular
      data analysis environment and programming language R, with a
      particular emphasis on comparative methods.

      2. The ability to parse and write NeXML from within R is only a
      small part of the functionality offered by the package.  We
      illustrate how RNeXML facilitates adding and manipulating data
      annotations whose meaning can be defined in a precise and
      machine-interpretable way by means of formal ontologies and other
      semantic web technologies.

      3. An essential feature of NeXML is to provide both the
      flexibility to be continually extended to meet the emerging
      needs of the phylogenetics community without breaking software
      unaware of such extensions. We illustrate how RNeXML makes
      it easy to extend NeXML in this manner.
      
      4. RNeXML is well suited for users to preserve and maintain
      their personal archives of phylogenetic and comparative trait
      data in consistent, query-able, and cross-platform format, to
      leverage the rapidly growing public community archives of rich
      NeXML files, and to interact with web services that return data
      and their metadata packaged together as NeXML.

bibliography: components/references.bib
csl: components/methods-in-ecology-and-evolution.csl
documentclass: components/elsarticle

output: 
  pdf_document:
    template: components/elsarticle.latex
    keep_tex: true
    fig_caption: true


---

<!------------------------------------------------

Map/Outline:
------------

- Introduction
  - Exciting times in phylogenetics!  
  - Challenges with existing formats 
  - Overview of NeXML
  - Challenges with R
  - Map 

- Tutorial
  - Reading Phylogenetic trees
  - Reading Character data
  - Writing phylogenetic trees & character data

  - Metadata
    - Adding basic metadata
    - Adding taxonomic identifiers
    - Adding Generic metadata
    - Parsing metadata. XML. RDF.
    - Extending NeXML with metadata: Example defining simmap stochastic character mappings.  

  - Publishing NeXML 

- Implementation, Quality Control
- Software reuse and support

------------------------------------------------>





```{r compile-settings, include=FALSE}
library("methods")
library("knitr")
opts_chunk$set(tidy = FALSE, warning = FALSE, message = FALSE, 
               cache = 1, comment = NA, verbose = TRUE)
basename <- gsub(".Rmd", "", knitr:::knit_concord$get('infile')) 
opts_chunk$set(fig.path = paste("components/figure/", basename, "-", sep=""),
               cache.path = paste("components/cache/", basename, "/", sep=""))


```

# Introduction

Users of the popular statistical and mathematical computing platform R
[@R] enjoy a wealth of readily installable comparative phylogenetic
methods and tools <!-- needs a citation to support it -->. Exploiting
the opportunities arising from this wealth for complex and integrative
comparative research questions relies on the ability to reuse and
integrate previously generated or published data and metadata. The
resulting expanding data exchange needs of the evolutionary research
community are rapidly outpacing the capabilities of the currently
still most widely used data exchange standards, which were all
developed a decade or more ago. This has resulted in a radiation of
different data representations and exchange standard "flavors" that
are no longer interoperable at the very time when the growth of
available data and methods has made that interoperability most
valuable. In response to the unmet needs for standardized data
exchange in phylogenetics, a modern XML-based exchange standard,
called NeXML, has recently been developed [@Vos_2012], and is
beginning to be adopted. NeXML comprehensively supports current data
exchange needs, is predictably machine-readable, and is forward
compatible. To make the capibilities of NeXML available to R users in
an easy-to-use form, and to lower the hurdles to adoption of the
standard, we present RNeXML, an R package that aims to provide easy
programmatic access to reading and writing NeXML documents, tailored
for the kinds of use-cases that will be common for users and
developers of the wealth of evolutionary, and in particular
comparative analysis methods within the R ecosystem.

The exchange problem for phylogenetic data is particularly acute in
light of the challenges in finding and sharing phylogenetic data
without the otherwise common loss of most data and metadata semantics
[@Drew_2013; @Stoltzfus_2012], or with semantics that is predictably
machine interpretable [@Cranston_2014]. For example, the still popular
NEXUS file format [@Maddison_1997] cannot consistently represent
horizontal gene transfer or ambiguity in reading a character (such as
a DNA sequence base pair).  This has led to modifications of NEXUS in
different ways for different needs, with the unfortunate result that
NEXUS files generated by one program can be incompatible with another
[@Vos_2012]. Without a formal grammar, software based on NEXUS files
may also make inconsistent assumptions about tokens, quoting, or
element lengths.  @Vos_2012 estimates that as many as 15% of the NEXUS
files in the CIPRES portal contain unrecoverable but hard to diagnose
errors.

A detailed account of how the NeXML standard addresses these and other
relevant challenges can be found in @Vos_2012. In brief, NeXML was
designed with the following important properties. First, NeXML is
defined by a precise grammar that can be programmatically
**validated**. This ensures that any NeXML file can be expected to be
parsed and read without errors by any NeXML-compliant software,
provided the file passes validation. Second, NeXML is **extensible**:
a user can define representations of new, previously unanticipated
information (as we will illustrate) without violating its defining
grammar. Third and most importantly, NeXML is **rich in computable
semantics**. The formal semantics used in NeXML is machine-readable in
a predictable manner, and also enables machine reasoning about the
data and metadata (see @Parr2011), especially when metadata attributes
and values are anchored in appropriately axiomatized ontologies.

`RNeXML` not only allows R's ecosystem to read and write data in
the NeXML format through an interface that is no more involved than
reading or writing data from other phylogenetic data formats, but also
carries immediate benefits for its users. For example, comparative
analysis R packages and users frequently add their own metadata
annotations to the phylogenies they work with, such as annotations
of species, stochastic character maps, trait values, model estimates
and parameter values. `RNeXML` affords R the capability to harness
machine-readable semantics and an extensible metadata schema to
capture, preserve, and share these and other kinds of information, all
through an API instead of having to understand in detail the schema
underlying the NeXML standard. To assist users in meeting the rising
bar for best practices in data sharing in phylogenetic research
[@Cranston_2014], `RNeXML` captures metadata information from the R
environment to the extent possible, and applies reasonable
defaults. 


# The RNeXML package

`RNeXML` is an R package entirely written in R, and the current stable
version can be installed from the CRAN library[^1]:

```r
install.packages("RNeXML", dependencies=TRUE)
```

This will also install its dependencies. Once installed, R scripts
using the library only need to load it:

```{r echo=FALSE}
library(RNeXML)
```

## Representation of NeXML documents in R

Conceptually, a NeXML document has the following components: (1)
phylogeny data, (2) character or trait data in matrix form, (3)
operational taxonomic units (OTUs), and (4) metadata. `RNeXML` defines
the `nexml` object type, which is a representation of the NeXML
document contents. It therefore holds phylogenetic trees as well as
character or trait matrices, and all metadata. This is similar in
design to the classes in the `phylobase` package, which contain trees
and traits but hold only limited metadata [@phylobase], but different
from R object types such as `ape::phylo` (the `phylo` class in the
`ape` package [@Paradis_2004]), which represents trees alone.

When reading and writing NeXML documents, `RNeXML` aims to map their
components to and from, respectively, their most natural
representations in R. Hence, the types of objects accepted or returned
by the package's methods are the `phylo` and `multiPhylo` objects from
the `ape` package [@Paradis_2004] for phylogenies, and R's native
`data.frame` list structure for character matrices.

## Reading phylogenies and character data

The method `nexml_read()` reads NeXML files, either from a local
file, or from a remote location via its URL, and returns an object of
type `nexml` [^2]:

```{r}
nex <- nexml_read("components/trees.xml")
```


To interface conveniently with the many already existing R packages for 
phylogenetics, which conventionally accept phylogenies as `ape::phylo`
objects, the method `get_trees_list()` returns a `ape::multiPhylo`
object, which can be treated as a list of `ape::phylo` objects:

```{r}
phy <- get_trees_list(nex)
```

The `get_trees_list()` method is designed for use in scripts, for
which a consistent and predictable return type (namely a `list` of
`ape::multiPhylo` objects, reflecting the nested structure of NeXML)
is desirable. `RNeXML` also contains a method `get_trees()`, which
is designed for interactive use and returns the R object most
intuitive given the arrangement of phylogeny data in the source NeXML
document. NeXML inherits from NEXUS the ability to contain multiple
phylogenies, and to group sets of phylogenies into "blocks". The
`get_trees()` method will return an `ape::phylo` object if the NeXML
document contains a single phylogeny, an `ape::multiPhylo` object if
it contains multiple phylogenies arranged in a single `trees` block,
and a list of `ape::multiPhylo` objects if it contains multiple
`trees` blocks. This design spares an interactive user from having to
type unnecessary data conversion commands, at the expense of needing
to know which return type to expect (or verifying the type via
introspection commands such as `class()`).

If the location parameter with which the `nexml_read()` method is
invoked is recognized as a URL, the method will automatically download the
document to the local working directory and read it from there. This
gives convenient and rapid access to phylogenetic data published in
NeXML format on the web, such as the content of the
phylogenetic data repository TreeBASE [@Piel_2002]. For
example, the following plots a tree in TreeBASE:

```{r eval=FALSE}
tb_nex <- nexml_read(
"https://raw.github.com/TreeBASE/supertreebase/master/data/treebase/S100.xml")
tb_phy <- get_trees_list(tb_nex)
```

The method `get_characters()` obtains character data matrics from a
`nexml` object, and returns them as a standard `data.frame` R object
with columns as characters and rows as taxa:

```{r }
nex <- nexml_read("components/comp_analysis.xml")
get_characters(nex)
```

A NeXML data matrix can be of molecular (for molecular sequence
alignments), discrete (for most morphological character data), or
continuous type (for many trait data).  Data matrices can be, and
matrices of different data types are required to be separated into
different "blocks", to allow strict validation of data types. Since
the `data.frame` data structure in R has no such constraints, the
`get_characters()` method will combine such blocks as separate columns
into a single `data.frame` object, provided they correspond to the
same taxa. Otherwise, a list of `data.frame`s is returned, with list
elements corresponding to characters blocks. Similar to the methods
for obtaining trees, there is also a method `get_characters_list()`,
which always returns a list of `data.frame`s, one for each character
block.

## Writing phylogenies and character data

The method `nexml_write()` generates a NeXML file from its input parameters. In
its simplest invocation, the method writes a tree to a file:

```{r}
data(bird.orders)
nexml_write(bird.orders, file = "birds.xml")
```

The first argument to `nexml_write()` is an object of type `nexml`, or any
object that can be coerced to it, such as in the above example an
`ape::phylo` phylogeny. Alternatively, passing a `multiPhylo` object
would write a list of phylogenies to the file.

In addition to trees, the `nexml_write()` method also allows to specify
character data as another parameter. The following example uses data
from the comparative phylogenetics R package `geiger` [@Harmon_2008].

```{r}
library("geiger")
data(geospiza)
nexml_write(trees = geospiza$phy, 
            characters = geospiza$dat,
            file="geospiza.xml")
```

## Creating and populating `nexml` objects

Instead of packaging the various components for a NeXML file at the
time of writing the file, `RNeXML` also allows users to create and
iteratively populate an in-memory `nexml` object. The methods to do
this are `add_characters()`, `add_trees()`, and `add_meta()`, for
adding characters, trees, and metadata, respectively. Each of these
functions will automatically create a new nexml object if not supplied
with an existing one as the last (optional) argument.

For example, here we use `add_trees()` to first create a `nexml`
object with the phylogeny data, and then add the character data to it:

```{r}
nexObj <- add_trees(geospiza$phy)
nexObj <- add_characters(geospiza$dat, nexObj)
```

The data with which a `nexml` object is populated need not share the
same OTUs.  `RNeXML` automatically adds new, separate OTU blocks into
the NeXML file for each data matrix and tree that uses a set of OTUs
that differs from others.

Other than storage size, there is no limit to the number of
phylogenies and character matrices that can be included in a
single NeXML file. This allows, for example, to capture samples from a
posterior probability distribution of inferred or simulated phylogenies or
character states in one NeXML file.

## Data documentation and annotation with built-in metadata

NeXML allows attaching metadata ("_annotations_") to any data element,
and even to metadata themselves. Whether at the level of the document
as a whole or the an individual data matrix or phylogeny, metadata can
provide bibliographic and provenance information, for example about
the study as part of which the phylogeny was generated or applied,
which data matrix and which methods were used to generate it. Metadata
can also be attached to very specific elements of the data, such as
specific traits, individual OTUs, nodes, or even edges of the
phylogeny.

NeXML uses the "Resource Description Framework in Annotations" (RDFa),
as described in @Vos_2012 or @W3C_2014, to encode metadata
annotations. This standard provides a precisely specified,
machine-readable format that maintains backwards compatibility with
compliant NeXML parsers (and thus `RNeXML`), because the capacity of a
tool to _parse_ annotations is not predicated on _understanding_
the meaning of annotations it has not seen before.

The burden typically involved with documenting and sharing good
metadata alongside the data remains a significant barrier to increased
sharing and reuse of scientific research [@Tenopir_2011]. To promote
sharing of well-documented phylogenetic data, `RNeXML` aims to
alleviate the burden to recording useful and machine-readable metadata
at several levels.

First, when writing a NeXML file the package adds certain basic
metadata automatically if it has not yet been provided, using default
values consistent with recommended best practices
[@Cranston_2014]. Currently, this includes naming the software
generating the NeXML, a time-stamp of when a tree was produced, and an
open data license.

Second, `RNeXML` provides a simple method, called
`add_basic_metadata()`, to set metadata attributes commonly
recommended to be included with publicly archived or shared data
[@Cranston_2014]. The currently accepted parameters include `title`,
`description`, `creator`, `pubdate`, `rights`, `publisher`, and
`citation`. Behind the scenes the method automatically anchors these
attributes in common vocabularies (such as Dublin Core). 

Third, `RNeXML` includes methods to assist with common cases of adding
metadata to specific elements of the NeXML data that increase the
data's utility for reuse. For example, misspellings and inconsistent
taxonomic names are a frequent source of frustration especially when
recombining phylogenetic data from different sources. These obstacles
can be mitigated by annotating the taxonomic names with identifiers,
which also allows being explicit about the identifier authority being
used. `RNeXML` includes the `taxize_nexml()` method, which uses the R
package `taxize` [@Chamberlain_2013] to match each taxon label against
the NCBI database, and, provided a unique match is found, annotates
the respective OTU with the matching NCBI identifier,.


## Data annotation with custom metadata

The RDFa standard adopted by NeXML (see @Vos_2012) for all metadata
annotations defines the structure and syntax for such annotations, but
does not restrict the content. Specifically, every metadata annotation
consists of a metadata property and a value of the property; the
bearer of the property is determined by where the annotation appears
in the NeXML document (technically, the containing element in the XML
document model). Technically, this structure corresponds directly to
Resource Description Framework (RDF) subject-predicate-object triples.
It therefore allows for arbitrary custom extensions to meet ongoing
scientific needs without requiring modifications (or "extensions") of
the syntax, provided the metadata annotations are cast as RDF triples.

The `RNeXML` methods described above for built-in metadata annotations
perform this cast automatically behind the scenes, including mapping
metadata attributes to terms in requisite common vocabularies (such as
Dublin Core for "title", "creator", etc.) or ontologies. Giving
metadata properties (and sometimes values) as vocabulary or ontology
terms rather than simple text strings is crucial for allowing machines
to not only parse but also interpret and potentially reason over their
semantics. To achieve this benefit for custom metadata extensions, the
user necessarily needs to handle certain technical details from which
the `RNeXML` API shields her otherwise, in particular the vocabularies
or ontologies of metadata terms in the form of globally unique
identifiers (normally HTTP URIs). To be consistent with XML
terminology for more technical users, `RNeXML` calls these
_namespaces_, and their abbreviations _prefixes_. For example, the
namespace for the Dublin Core Metadata Terms vocabulary, commonly
abbreviated with "dc", is "http://purl.org/dc/elements/1.1/". Using
its abbreviation, a metadata property "dc:title" expands to the
identifier "http://purl.org/dc/elements/1.1/title", the identifier for
the term "title" in said vocabulary. When resolved as a HTTP URI, the
identifier resolves to a human and machine-readable
(depending on how the identifier is resolved) definition of precisely
what the term `title` in Dublin Core means. In contrast, just using
the text string "title" could also mean the title of a person, a legal title, the verb title, etc.  Although URI identifiers of metadata
vocabularies and terms are not required to resolve, for machines to
derive the maximum benefit from them, they should resolve to RDF
statements defining their semantics.

`RNeXML` includes methods to obtain and manipulate metadata
properties, values, identifiers, and namespaces. The
`get_namespaces()` method accepts a `nexml` object and returns a list
with element names being the namespace abbreviations and element
values being the corresponding identifiers (typically HTTP URIs) for
all annotations attached to the object or one of the contained data elements:


```{r}
prefixes <- get_namespaces(birds)
prefixes["dc"]
```

The `get_metadata()` method returns, as named list, the metadata
annotations for a given `nexml` object at a given level, with the
whole NeXML document being the default:

```{r}
meta <- get_metadata(birds) 
```

The `level` parameter can be used to extract the metadata objects for
the data at specific levels (such as `"otu"` in the following
example), or by specifying `"all"` to extract all metadata objects:

```{r} 
otu_meta <- get_metadata(birds, level="otu")
```

However, because the returned list does not include the data elements
to which the metadata are attached, using the method in this way is
only useful for identifying which metadata are present and their
values. The recommended approach for accessing the metadata attached
to specific data elements is described below. <!-- or is that in the vignette now? -->

A few select metadata properties for which support is built-in have
dedicated accessor methods, for example the citation:

```{r get_citation}
get_citation(birds)
```

The `meta()` method creates a new metadata object from a poperty name
and content (value). For example, the following creates a modification
date metadata object, using the modificationDate property in the PRISM
vocabulary:


```{r}
modified <- meta(property = "prism:modificationDate", content = "2013-10-04")
```

The `add_meta()` function adds metadata objects as annotations to a
`nexml` object at a specified level. The default level is the NeXML document as a whole:

```{r}
birds <- add_meta(modified, birds) 
```

If the prefix used by the metadata property is not among the built-in
ones (which can be obtained using the `get_namespaces()` as described
above), it has to be provided along with its URI as the `namespaces`
parameter. For example, the following uses the SKOS vocabulary (which
currently is not built-in) to add a note to the trees in the `nexml`
object:

```{r}
history <- meta(property = "skos:historyNote",
  content = "Mapped from the bird.orders data in the ape package using RNeXML")
birds <- add_meta(history, 
                birds, 
                level = "trees",
                namespaces = c(skos = "http://www.w3.org/2004/02/skos/core#"))
```


`RNeXML`, by virtue of subsetting the S4 `nexml` object, also offers
fine control of the level at which a `meta` element is added, for
which the package vignette contains examples.

Because NeXML expresses all metadata using the RDF standard, and
stores them compliant with RDFa, they can be extracted as an RDF graph
using off-the-shelf tools, and the RDF graph obtained in this way can
be queried in SPARQL (see @W3C_2014 or @Hartig_2012), the query
language for RDF graphs, using general purpose triple stores and
SPARQL query engines. The `RNeXML` vignette illustrates this with
examples, and the package also comes with a demonstration that can be
run from R using the following command: `demo("sparql", "RNeXML")`).

## Using metadata to extend the NeXML standard 

No data standard can anticipate all the uses and extensions
researchers may one day need. As the various "flavors" of the NEXUS
and other data formats show, this often results in extensions to
widely used exchange formats that almost invariably break
interoperability between such extensions and tools that were designed
to read one but not another. NeXML was designed to prevent this by
means of a highly flexible metadata system, while not sacrificing 
strict validation of syntax and structure.

Here we illustrate how NeXML's metadata system, and `RNeXML`'s API to
it, can be used to record and share a type of phylogenetic data not
taken into account when NeXML was designed. Specifically, we use the
example of stochastic character maps [@Huelsenbeck_2003], which assign
sections of each branch in a phylogeny to a particular "state"
(typically of a morphological characteristic). The current de-facto
format for sharing stochastic character maps is a (non-interoperable)
modification of the standard Newick tree format created by a software
tool widely used for creating such maps, `simmap` [@Bollback_2006].

@Revell_2012 describes the `phytools` package for R, which includes
utilities for reading, manipulating, and writing `simmap` files in R.
In this example, we also show how to define `RNeXML` functions that 
map the R representation used by Revell (an extension of the `ape` class)
into the NeXML extension we have defined by using `RNeXML` functions.

Since a stochastic character map simply assigns different states to
parts of a branch (or edge) on the phylogenetic tree, we can create
a NeXML representation by annotating the `edge` elements with appropriate
`meta` elements.  These elements need to describe the character state
being assigned and the duration (in terms of branch-length) that the edge 
spends in that state (Stochastic character maps are specific to time-calibrated
or ultrametric trees).  

NeXML already defines the `characters` element to handle discrete character traits (`nex:char`)
and the states they can assume (`nex:state`).  We will thus reuse the `characters` element for
this purpose, referring to both the character trait and the states by the ids assigned to them
in that element.  (NeXML's convention of referring to everything by id permits a single canonical
definition of each term, making it clear where additional annotation belongs).  For each edge, we 
need to indicate:

<!-- Ugh, do we need to explain that everything in NeXML uses an id so that it can be reused in this way?
     -->

- That our annotation contains a stochastic character mapping reconstruction
- Since many reconstructions are possible for a single edge, we give each reconstruction an id
- We indicate for which character trait we are defining the reconstruction 
- We then indicate which states the character assumes on that edge. 
  For each state realized on the edge, that involves stating: 
    + the state assignment
    + the duration (length of time) for which the edge spends in the given state
    + the order in which the state changes happen (Though we could just assume 
      state transitions are listed chronologically, NeXML suggests making all 
      data explicit, rather than relying on the structure of the data file to
      convey information).  

Thus the annotation for an edge that switches from state `s2` to state 
`s1` of character `cr1` would be constructed like this:

```{r}
 m <- meta("simmap:reconstructions", children = c(
        meta("simmap:reconstruction", children = c(

          meta("simmap:char", "cr1"),
          meta("simmap:stateChange", children = c(
            meta("simmap:order", 1),
            meta("simmap:length", "0.2030"),
            meta("simmap:state", "s2"))),
          
          meta("simmap:char", "cr1"),
          meta("simmap:stateChange", children = c(
            meta("simmap:order", 2),
            meta("simmap:length", "0.0022"),
            meta("simmap:state", "s1")))
          ))))
```

Of course writing out such a definition manually becomes tedious quickly. Because
these are just R commands, we can easily define a function that can loop over an
assignment like this for each edge, extracting the appropriate order, length and
state from an existing R object such as that provided in the `phytools` package.  
Likewise, it is straightforward to define a function that reads this data using
the `RNeXML` utilities and converts it back to the `phytools` package. The full
implementation of this mapping can be seen in the `simmap_to_nexml()` and the
`nexml_to_simmap()` functions provided in the `RNeXML` package.  

As the code indicates, the key step is simply to define the data in meta elements. In 
so doing, we have defined a custom namespace, `simmap`, to hold our variables.  This
allows us to provide a URL with more detailed descriptions of what each of these 
elements mean:

```{r}
nex <- add_namespaces(c(simmap = "https://github.com/ropensci/RNeXML/tree/master/inst/simmap.md"))
```

At that URL we have posted a simple description of each term. 

Using this convention we can generate NeXML files containing `simmap`
data, read those files into R, and convert them back into the `phytools`
package format. These simple functions serve as further illustration of
how `RNeXML` can be used to extend the NeXML standard.  We illustrate
their use briefly here, starting with  loading a `nexml` object containing
a `simmap` reconstruction into R:


```{r}
data(simmap_ex)
```

The `get_trees()` function can be used to return an `ape::phylo` tree as
usual.  `RNeXML` automatically detects the `simmap` reconstruction data
and returns includes this in a `maps` element of the `ape::phylo` object,
for use with other `phytools` functions.

```{r}
phy <- nexml_to_simmap(simmap_ex)
```

We can then use various functions from `phytools` designed for `simmap`
objects [@Revell_2012], such as the plotting function:

```{r Figure1, fig.cap="Stochastic character mapping on a phylogeny, as generated by the phytools package after parsing the simmap-extended NeXML."}
library("phytools")
plotSimmap(phy)
```

Likewise, we can convert the object back in the NeXML format and write
it out to file to be read by other users. 

```{r}
nex <- simmap_to_nexml(phy) 
nexml_write(nex, "simmap.xml")
```

Though other NeXML parsers (for instance, for Perl or Python) have
not been written explicitly to express `simmap` data, those parsers will
nonetheless be able to successfully parse this file and expose the `simmap`
data to the user.





## Publishing NeXML files from R

Data archiving is increasingly required by journals in evolutionary biology and 
biodiversity (e.g. @Rausher_2010).  The burden of preparing and submitting properly
annotated data to archives continues to be a significant barrier [@Tenopir_2011; @Stodden_2014],
and many phylogenetic trees upon which studies are based are inaccessible or lost 
to the research community [@Drew_2013].  `RNeXML` seeks to lower these barriers
by providing functions to immediately archive NeXML files in scientific repositories.  




```{r eval = FALSE}
doi <- nexml_publish(birds, visibility = "public", repository="figshare")
```

Robust data archiving need not involve immediate release of the data. The `figshare` 
repository supports the option of `visibility = "private"`, allowing secure 
publication that provides a way to backup and share data with collaborators 
prior to public release.  


# Software reuse and support

The `RNeXML` package is available through the CRAN repository under
a CC0 license. At the time of writing, the current version is `r packageDescription("RNeXML", fields="Version")`. 
The current development
branch can be found on Github at [https://github.com/ropensci/RNeXML]().
See the NEWS file for changes in recent versions. Users are encouraged
to submit bug reports or feature requests in the issues log.  Contact
`ropensci-discuss@googlegroups.com` or the package maintainer for help,
or email the phylogenetics R users group list at `r-sig-phylo@r-project.org`.  


There are many promising directions for further development of the
`RNeXML` package.  The project's [Milestones page](https://github.com/ropensci/RNeXML/issues/milestones) page provides a list of
steps currently planned or under consideration.  In particular, these
milestones include increased support for interactive mode of entering data
(wizards), conversions to and from other existing R phylogenetics formats
such those found in the `phylobase` package [@phylobase],
and support for integration of semantics defined through the
Web Ontology language (OWL), as discussed in the Challenges section.

## Future Directions

For convenience, the RNeXML package can parse compressed XML binaries (gzipped XML) directly, facilitating work with very large data files.  As most R functions, especially those in common phylogenetics packages, require in-memory data structures, RNeXML currently loads this data directly into memory. Future work could facilitate manipulation of NeXML data directly from the disk.

Emerging formats such as the JSON based NeXSON format provide alternate representations of semantic, machine-readable phylogenetic data. At this time, R users must still rely on other tools that can convert these formats into NeXML.
Because the NeXML syntax is extensible, semantic, and machine-readable, this can be done without loss, unlike NEXUS formats.  Nevertheless, a possible future direction would be to provide direct support of such formats through RNeXML or a companion R package.

## Acknowledgements
<!-- FIXME Rutger, Hilmar, add whatever you'd like to acknowledge here -->. 

This project was supported in part by a grant from the Alfred P Sloan
Foundation Grant 2013-6-22 (CB and SC), NSF Grant DBI-1306697 (CB), ... 
We thank F. Michonneau for helpful comments on an earlier version
of this manuscript.
`RNeXML` is part of the rOpenSci project, [http://ropensci.org].



```{r cleanup, include=FALSE, cache=FALSE}
unlink("simmap.xml")
unlink("birds.xml")
unlink("geospiza.xml")
unlink("rdf_meta.xml")
```


<!-- Footnotes -->

[^1]: By providing the optional `dependencies` argument, R will install
additional packages used in some of the advanced examples shown later
but which are not required for basic functionality.

[^2]: For convenience `nexml_read` is also aliased as `read.nexml`,
for users familiar with the somewhat analogous `read.nexus` function.

[^3]: By using the commenting mechanism of the Newick format, it is
possible that other software that doesn't also use the comment mechanism
for some other such purpose would be able to successfully parse the tree.
However there is no way to guarantee that this is the case or for the
data format to describe its use.







