
## Abstract

Why RNeXML?

* Why NeXML -- because Nexus standard isn't a standard.  Developers extend it arbitrarily to meet their needs, making it incompatible.  Cannot represent horizontal gene transfer, ambiguous use of symbols, cannot be extended reliably, no validator.  NeXML solves all of these problems [Vos _et al._ 2012] with a state-of-the-art data management format.  

* Why RNeXML: Huge ecosystem of phylogenetics packages. (1) Read nexml.  All would benefit by being able to read NeXML, both as a robust standard free of the pitfalls of Nexus, and more importantly, as a much richer source of data thanks to its extendable annotation model.  (2) Write nexml.  R packages and R users add rich information to the phylogenies they work with, but need a platform independent data exchange format to share and archive this. (3) Annotate and extend nexml.  

* as your own personal treebase

* Adding metadata to capture workflow


## Background

### Challenges with existing formats
See [Vos _et al._ 2012].  

### Challenges with R formats

See [phylobase vignette, 2010]

### NeXML

See [Vos _et al._ 2012].

Mention of data sharing challenges and opportunities.  See [Drew et al 2013](10.1371/journal.pbio.1001636)



A nexml file can be thought of as having three components: (1) phylogeny data (2) character-trait data, and (3) metadata. `RNeXML` seeks to provide a seemless way to convert these to and from their corresponding R object representations, such as `ape::phylo` phylogenies, character matrices (`data.frames`) and metadata lists (or for advanced users, `XMLInternalDocument` and `rrdf` triplestores).  The RNeXML package functions (the package Application Programming Interface, API) 

## Introduction to RNeXML

<!-- Do we describe installation of the package?  `install.packages("RNeXML")`?-->


## Working with phylogenetic trees

#### Read methods

```{r echo=FALSE}
library(RNeXML)
```

We begin by reading in an example NeXML file.   

```{r}
f <- system.file("examples", "trees.xml", package="RNeXML")
nex <- nexml_read("trees.xml")
```

(For convenience `nexml_read` is also aliased as `read.nexml`, for users familiar with the somewhat analogous `read.nexus` function).   The resulting `nex` object is an R object representation of the NeXML.  From this object we can extract any phylogenies it contains in the `ape::phylo` format:   

```{r}
phy <- get_trees(nex)
```
We can then leverage the rich suite of methods available from the R phylogenetics community (including coercion into other representations, such as `ouch` [@ouch] and `phylobase` [@phylobase] phylogeny formats). For instance, here we use the standard plotting function from the `ape` package to plot the resulting `ape::phylo` object returned by `get_trees`.  


[@ouch]: http://cran.r-project.org/web/packages/ouch/
[@phylobase]: http://cran.r-project.org/web/packages/phylobase

``{r}
plot(phy)
```

NeXML can not only contain multiple phylogenetic trees in a single file, but can also group sets of trees into separate blocks.  The `get_trees` function attempts to return the most sensible R object for NeXML: if only a single tree is available, it will return an `ape::phylo` object; if multiple trees are available in a single `trees` block, it will return an `ape::multiPhylo` object (a list of `phylo` objects).  If multiple blocks are available, it will return a list of `multiPhylo` blocks.  While most users will find this behavior convenient, there are times (such as looping over many NeXML files) when having a variable return type is undisirable.  To always get back a list of `multiPhylo` objects, even when only one tree is available, use `get_trees_list()`.  

<!-- have we already said all this above?-->
The ability to read in phylogenetic trees in the `nexml` format opens up a wide and rapidly growing array of resources to the R user.  For instance, a user can access all the phylogenies available in TreeBASE through the `nexml` format.  `RNeXML` can read directly from a URL: 

```{r eval=FALSE}
nex = nexml_read("https://raw.github.com/rvosa/supertreebase/master/data/treebase/S100.xml")
```

Previously, this was possible using the `treebase` package [@treebase] for R, which relied on the nexus parser provided in the `ape` package.  Due to inconsistencies present in the Nexus standard [Vos _et al._ 2012], a significant number <!-- State explicitly --> of TreeBASE nexus files could not be read in this way.  Using the more precisely defined NEXML standard, every valid NeXML file on TreeBASE can now successfully be parsed.  




#### Write methods 








##### Character data


### Reading NeXML metadata

Replicate examples from [treebase] metadata section: publication information, tree size, taxanomic coverage

### Writing NeXML phylogenies

```{r}
write.nexml(phy, "example.xml")
```

### Reading NeXML metadata

One of the most powerful features of the NeXML format is the ability add arbitrary metadata information using the RDFa annotation standard [cite W3C]().   

Metadata from treebase NeXML
- article citation information
- Summary data: taxa, number of taxa, number of trees, number of tips.  
- Metadata of nodes, complete RDF extraction

Using the nexml and nexmlTree class for metadata.  


### Writing NeXML metadata

`RNeXML` also helps users add richer metadata when producing NeXML files. This occurs on several levels:

- Automatic annotation.  `RNeXML` automatically adds basic metadata by default such as a timestamp of when the tree was produced and an open data license (CC0, compatible with deposition on Dryad data repository), which can be configured as needed.  

- Utilities to add common metadata

  - Add a citation to any papers associated with the phylogeny data.  
  - Writing in methodological information on how tree has been generated or analyzed (including R code)
  - Add annotations to any taxon.  

Writing annotations on otus.  Annotations on nodes.  

Temporal, spatial, taxonomic coverage for dryad/knb?

### Implementation and architecture

The `RNeXML` package is built for the NeXML Schema, [version 2009], using the R software environment and the XML package for parsing and generating XML [Temple Lang 2013, XML], and provides conversions into the popular `phylo` format provided by the `ape` package [Paradis 2004].  `RNeXML` can automatically check taxonomic names and add taxonomic identifiers when generating NeXML using methods from the `taxize` package ([Chamberlain 2013]). 

<!-- Possibly cite all dependencies and suggests from the DESCRIPTION --> 
<!-- Reference any installation and configuration instructions necessary (e.g. API credentials for figshare) -->

[Temple Lang 2013, XML]: http://cran.r-project.org/web/packages/XML/

### Quality control

The `RNeXML` package performs two levels of validation on NeXML files written from or read into R: validating the XML, and validating against the current NeXML schema.

Several mechanisms are in place to evaluate the quality of the `RNeXML` package functions themselves.  In addition to the examples from the documentation and the automated package checks provided for the R system, the `RNeXML` package includes a complete test suite using the testthat package [Wickham 2011].  

[Wickham 2011]: http://vita.had.co.nz/papers/testthat.html

### Software reuse and support

The `RNeXML` package is available through the CRAN repository under a CC0 license. At the time of writing, the current version is `r packageDescription("RNeXML", fields="Version")`. The current development branch can be found on Github at [https://github.com/ropensci/RNeXML].  See the NEWS file for changes in recent versions. Users are encouraged to submit bug reports or feature requests in the issues log.  Contact ropensci-discuss@googlegroups.com or the package maintainer for help. 

There are many promising directions for further development of the `RNeXML` package.  The project's Milestones page provides a list of steps currently planned or under consideration.  In particular, these milestones include increased support for interactive mode of entering data (wizards), and support for integration of semantics defined through the Web Ontology language (OWL), as discussed in the Challenges section.  



### Acknowledgements

This project was supported in part by a grant from the Alfred P Sloan Foundation (CB and KR), NSF Grant DBI-1306697 (CB), ... `RNeXML` is part of the rOpenSci project, [http://ropensci.org].  





phylobase::phylo4 was an attempt to solve the problems associated with the popular ape::phylo format (unpredictable errors, loose and multiple conflicting specifications).  



